
Raised as PR upstream - https://github.com/cyrusimap/cyrus-imapd/pull/3174

commit 98bf19e382b0161386adece9ff9a1ba07ed0bf60
Author: Andy Fiddaman <omnios@citrus-it.co.uk>
Date:   Sat Sep 5 15:06:29 2020 +0000

    Not all systems have futimes() or TIMESPEC_TO_TIMEVAL()

diff -wpruN '--exclude=*.orig' a~/configure.ac a/configure.ac
--- a~/configure.ac	1970-01-01 00:00:00
+++ a/configure.ac	1970-01-01 00:00:00
@@ -194,7 +194,7 @@ CMU_GUESS_RUNPATH_SWITCH
 
 AC_CHECK_HEADERS(unistd.h sys/select.h sys/param.h stdarg.h)
 AC_REPLACE_FUNCS(memmove strcasecmp ftruncate strerror posix_fadvise strsep memmem)
-AC_CHECK_FUNCS(strlcat strlcpy strnchr getgrouplist fmemopen pselect)
+AC_CHECK_FUNCS(strlcat strlcpy strnchr getgrouplist fmemopen pselect futimes futimesat)
 AC_HEADER_DIRENT
 
 dnl check whether to use getpassphrase or getpass
diff -wpruN '--exclude=*.orig' a~/lib/util.c a/lib/util.c
--- a~/lib/util.c	1970-01-01 00:00:00
+++ a/lib/util.c	1970-01-01 00:00:00
@@ -574,7 +574,13 @@ static int _copyfile_helper(const char *
         struct timeval tv[2];
         TIMESPEC_TO_TIMEVAL(&tv[0], &sbuf.st_atim);
         TIMESPEC_TO_TIMEVAL(&tv[1], &sbuf.st_mtim);
+#if defined(HAVE_FUTIMES)
         if (futimes(destfd, tv)) {
+#elif defined(HAVE_FUTIMESAT)
+	if (futimesat(destfd, NULL, tv)) {
+#else
+	if (utimes(to, tv)) {
+#endif
             syslog(LOG_ERR, "IOERROR: setting times on %s: %m", to);
             r = -1;
         }
diff -wpruN '--exclude=*.orig' a~/lib/util.h a/lib/util.h
--- a~/lib/util.h	1970-01-01 00:00:00
+++ a/lib/util.h	1970-01-01 00:00:00
@@ -148,6 +148,13 @@ extern const unsigned char convert_to_up
 /* Calculate the number of entries in a vector */
 #define VECTOR_SIZE(vector) (sizeof(vector)/sizeof(vector[0]))
 
+#ifndef TIMESPEC_TO_TIMEVAL
+#define TIMESPEC_TO_TIMEVAL(tv, ts) { \
+        (tv)->tv_sec = (ts)->tv_sec; \
+        (tv)->tv_usec = (ts)->tv_nsec / 1000; \
+}
+#endif
+
 typedef struct keyvalue {
     char *key, *value;
 } keyvalue;
